### ISSUE TITLE
Question: What is the purpose of dist/immutable-nonambient.d.ts? 

### ISSUE THREAD

**Issue by KSXGitHub** [facebook/immutable-js#1662]
Why does it exist?
What happen if it is removed?

**Comment by conartist6** [comment]
According to https://github.com/Microsoft/TypeScript-Handbook/issues/180

> A module can be ambient as well; so
> "ambient module" is a .d.ts file with a top level import or export.

It exists to bridge the gap between CommonJS and ES6 modules. When used as a CommonJS module, Immutable replaces the entire exports object with its own, which is what Immutable.d.ts describes. In an es6 environment, it has named exports for each of its various classes and utility functions. The non-ambient types describe this method of using the library.

The one piece of the puzzle I'm still missing is how Immutable expects the non-ambient definitions to be picked up.

**Comment by conartist6** [comment]
Oh right, it's all in package.json:
```
  "typings": "dist/immutable-nonambient.d.ts",
  "typescript": {
    "definition": "dist/immutable.d.ts"
  },
```

**Comment by conartist6** [comment]
So to fully answer your question: if you remove it, people who use the syntax
```
import { List } from "immutable"
```
will be angry at you.

**Comment by KSXGitHub** [comment]
> So to fully answer your question: if you remove it, people who use the syntax
> 
> ```
> import { List } from "immutable"
> ```
> will be angry at you.

So what you are saying is TypeScript will pick `dist/immutable-nonambient.d.ts` when user uses ES6 import syntax and `dist/immutable.d.ts` for `require()` syntax?

I have take a look at the 2 files again, the only different I can see is everything in `dist/immutable.d.ts` is wrap within a namespace called `Immutable`. The "ambient" version exports that big namespace while the "non-ambient" export everything directly.

This leads me to another question: To my knowledge, `dist/immutable-nonambient.d.ts` should support `require()` syntax, does this make the "ambient" version redundant? Or in other words, `immutable.d.ts` should be written in "non-ambient" style in the first place.

**Comment by conartist6** [comment]
I think you are right on all counts.

I just did some experimental testing and using the `import = require ...` syntax and it seems that it exports an object, as I would expect. The only confusion would be a default key on the object if a default export was defined, but the nonambient types do not define a default export.

**Comment by conartist6** [comment]
By the way, I suspect the reason it's still there is that the way the API doc website works is that it basically is given an approximation of the AST of that file and then uses it to generate the site docs. Changing the shape of the AST like that would require a cascade of changes to get the doc site back to working, and the doc site already isn't quite working the way its meant to, and is untested.

**Comment by conartist6** [comment]
@KSXGitHub I forked the Immutable doc generator for my mutable immutable.js clone, called sequins. After this conversation I went ahead and made [the changes that you suggest](https://github.com/conartist6/sequins/commit/be5ff23541fcc95ad399009a98ba4a8722f7340e) in my project.

The only major difference in the fork for the files I touched is that I guarantee that all the types I export are at the top level (not namespaced), whereas Immutable has some, for example, `Seq.Keyed`. This allowed me to simplify the original logic around `type.qualifier`.
